materialize-sql: use `encrow.Shape` for encoding

williamhbaker opened the pull request:
**Description:**

Use the optimized `encrow.Shape` for encoding JSON objects, and also optimizations for encoding an array of values for materializations that do that (currently only `materialize-snowflake`).

This is a significant performance improvement when examining _just_ the serialization aspect of the `CountingEncoder`, and shows a 3x improvement in speed of encoding objects, and a ~1.5x improvement for serializing value arrays.

The benchmark outputs below illustrate this, and were run with `compressionLevel = flate.NoCompression` to turn off compression in `CountingEncoder`. The original pre-change code was tested with a slightly modified version of the benchmark included in this PR for compatibility.

_New optimized code, with `compressionLevel = flate.NoCompression`:_
```
go test -v ./materialize-sql/ -bench=. -benchtime=5s -count=3
BenchmarkEncodingObjects
BenchmarkEncodingObjects-10         6801            895722 ns/op
BenchmarkEncodingObjects-10         6858            903788 ns/op
BenchmarkEncodingObjects-10         6705            896007 ns/op
BenchmarkEncodingArrays
BenchmarkEncodingArrays-10          7338            837367 ns/op
BenchmarkEncodingArrays-10          7232            842599 ns/op
BenchmarkEncodingArrays-10          7318            835675 ns/op
```

_Original code, also with `compressionLevel = flate.NoCompression`:_
```
go test -v ./materialize-sql/ -bench=. -benchtime=5s -count=3
BenchmarkEncodingObjects
BenchmarkEncodingObjects-10         2540           2357578 ns/op
BenchmarkEncodingObjects-10         2544           2372622 ns/op
BenchmarkEncodingObjects-10         2546           2386773 ns/op
BenchmarkEncodingArrays
BenchmarkEncodingArrays-10          4599           1305916 ns/op
BenchmarkEncodingArrays-10          4588           1300114 ns/op
BenchmarkEncodingArrays-10          4659           1302175 ns/op
```

Of course, materialization connectors do more than just JSON serialization, and the `CountingEncoder` abstraction also does compression at the `flate.BestSpeed` level. Running the benchmarks with compression enabled shows a lesser benefit, indicating that the work of compression is starting to overshadow the work of JSON serialization. With compression enabled, `CountingEncoder` is about 40% faster at encoding objects, and 15% faster at encoding value arrays with these changes. This is still a decent boost, and I'm hopeful it will provide a noticeable improvement, particularly for materializations handling really large documents or with huge numbers of fields.

_New optimized code, with compression enabled:_
```
go test -v ./materialize-sql/ -bench=. -benchtime=5s -count=3
BenchmarkEncodingObjects
BenchmarkEncodingObjects-10         1660           3500257 ns/op
BenchmarkEncodingObjects-10         1707           3517238 ns/op
BenchmarkEncodingObjects-10         1700           3511234 ns/op
BenchmarkEncodingArrays
BenchmarkEncodingArrays-10          1582           3780298 ns/op
BenchmarkEncodingArrays-10          1580           3783878 ns/op
BenchmarkEncodingArrays-10          1569           3794444 ns/op
```

_Original code, with compression enabled:_
```
go test -v ./materialize-sql/ -bench=. -benchtime=5s -count=3
BenchmarkEncodingObjects
BenchmarkEncodingObjects-10         1213           4960839 ns/op
BenchmarkEncodingObjects-10         1204           4974379 ns/op
BenchmarkEncodingObjects-10         1168           5024522 ns/op
BenchmarkEncodingArrays
BenchmarkEncodingArrays-10          1393           4324955 ns/op
BenchmarkEncodingArrays-10          1407           4342299 ns/op
BenchmarkEncodingArrays-10          1386           4333667 ns/op
```


**Workflow steps:**

(How does one use this feature, and how has it changed)

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

This was tested using the integration tests for `materialize-bigquery`, `materialize-redshift`, and `materialize-snowflake`.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/estuary/connectors/1066)
<!-- Reviewable:end -->

mdibaiee commented on tests/materialize/materialize-redshift/snapshot.json:
```diff
@@ -645,6 +645,42 @@
+      "int_str": null,
+      "num_and_str": null,
+      "num_str": null,
+      "time": null
```
How did we end up with these new records in the snapshot? 🤔

williamhbaker replied:
From #1037, which added some new test cases, and must not have run the redshift test/updated the snapshots. I ran it for testing this PR and updated the snapshot, which brought in that change as well, although unrelated to anything here.

mdibaiee reviewed (approved):
LGTM, this will be great addition to databricks

willdonnelly reviewed (approved):
LGTM

williamhbaker commented:
One addendum: I'm swapping in `github.com/klauspost/compress/gzip` in place of the standard gzip package. This improves the encoding benchmarks for about 40% with compression enabled:

_New optimized code, with compression enabled using `github.com/klauspost/compress/gzip`:_
```
go test -v ./materialize-sql/ -bench=. -benchmem -benchtime=5s -count=3
BenchmarkEncodingObjects
BenchmarkEncodingObjects-10         2302           2528709 ns/op
BenchmarkEncodingObjects-10         2474           2435775 ns/op
BenchmarkEncodingObjects-10         2461           2417468 ns/op
BenchmarkEncodingArrays
BenchmarkEncodingArrays-10          2269           2667515 ns/op
BenchmarkEncodingArrays-10          2280           2653695 ns/op
BenchmarkEncodingArrays-10          2258           2660670 ns/op
```

williamhbaker commented:
Another addendum: I had removed the `countingWriteCloser` proxy struct thinking it was a nice simplification, but it was actually necessary to output files at the target file size of _compressed_ bytes, so I added that back in.
